src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1120,41 +1120,65 @@
-    "                            when 'array'"
-    "                              then generate_series(0, jsonb_array_length(value - 1))::text"
-    "                            end"
-    "                            as sub_path,"
-    "                          case jsonb_typeof(value)"
-    "                            when 'object'"
-    "                              then (jsonb_each(value)).value"
-    "                            when 'array'"
-    "                              then jsonb_array_elements(value)"
-    "                          end"
-    "                          as sub_value"
-    "                     from live_paths) as candidates"
-    "             where candidates.sub_path is not null)"
-    "               as sub_paths)"
-    "  delete from fact_paths fp"
-    "    where not exists (select 1 from live_paths"
-    "                        where live_paths.path = fp.path)"]))
+   ["with recursive live_paths(path, value) as"
+    "   (select key as path, value"
+    "      from (select (jsonb_each(stable||volatile)).* from factsets) as base_case"
+    "      union"
+    "        select path||'#~'||sub_level.key as path,"
+    "               sub_level.value"
+    "          from live_paths,"
+    "          lateral (select *"
+    "                     from (select (jsonb_each(value)).*"
+    "                             where jsonb_typeof(value) = 'object') as sub_fields"
+    "                     union (select generate_series(0, jsonb_array_length(value - 1))::text as key,"
+    "                                   jsonb_array_elements(value) as value"
+    "                              where jsonb_typeof(value) = 'array')) as sub_level)"
+    "   delete from fact_paths fp"
+    "     where not exists (select 1 from live_paths"
+    "                         where live_paths.path = fp.path)"]))
+
+(defn delete-unused-fact-paths-via-temp-table []
+  "Deletes paths from fact_paths that are no longer needed by any
+  factset.  In the unusual case where a path changes type, the
+  previous version will linger."
+  (jdbc/do-commands
+   ["with recursive live_paths(path, value) as"
+    "   (select key as path, value"
+    "      from (select (jsonb_each(stable||volatile)).* from factsets) as base_case"
+    "      union"
+    "        select path||'#~'||sub_level.key as path,"
+    "               sub_level.value"
+    "          from live_paths,"
+    "          lateral (select *"
+    "                     from (select (jsonb_each(value)).*"
+    "                             where jsonb_typeof(value) = 'object') as sub_fields"
+    "                     union (select generate_series(0, jsonb_array_length(value - 1))::text as key,"
+    "                                   jsonb_array_elements(value) as value"
+    "                              where jsonb_typeof(value) = 'array')) as sub_level)"
+    "   select path into temp tmp_live_paths from live_paths"]
```

Comment:
Just to make sure I'm not missing something, it looks like we're throwing the results of the cte above into the tmp_live_paths table and then using that table to compare which facts are still live in the real table deleting those which don't exist in the tmp table, is that right? If that is the case it seems like we're performing the deletion step in roughly the same way in both functions, was this what we wanted to do? I remember us talking about possibly creating a table that was the result of the live paths query and just keeping that live_paths table while dropping the old table wholesale.

Reply:
Yeah, that's what the current code intends, i.e. they do the same thing, but using the temp table hopefully avoids the tiny temp file log flood.